File: Sources/Applicazione/main/java/it/kaizenteam/app/model/NorrisChart/MapChartInPlaceUpdater.java
/*
* Name: MapChartInPlaceUpdater.java
* Package: it.kaizenteam.app.model.NorrisChart
* Location: Sources/Applicazione/main/java/it/kaizenteam/app/model/NorrisChart
* Date: 2015-05-19
* Version: v0.02
*
* History:
* =================================================================
* Version	Date	Programmer	Changes
* =================================================================
* v0.02 2015-05-24  Davide Dal Bianco   Verify
* =================================================================
* v0.01 2015-05-18  Moretto Alessandro  Creazione file
* =================================================================
*
*/

package it.kaizenteam.app.model.NorrisChart;

/**
 * That class is responsible for defining in place updating method for a map chart. It can access to the private data fields of MapChartImpl because it is an inner class. In particular it can access to DataObject contained in MapChartImpl and change its values.
 */
public class MapChartInPlaceUpdater implements Updater {
    /**
      * The static attribute represent the unique instance of the class. 
      */
    private static MapChartInPlaceUpdater instance;

    /**
     * This method has the task of returning the unique instance of class, and creating it if it not exists.
     * @return instance of the class
     */
    public static Updater getInstance(){
        return instance;
    }

    /**
     * Constructor of the class
     */
    private MapChartInPlaceUpdater(){}


    /**
     * This method has the task of updating the chart passed as a parameter (chart) by using the update package (UpdateData).
     * @param chart the chart to be update
     * @param updateData the update package
     */
    @Override
    public void update(ChartImpl chart, ChartUpdate updateData) {
//TODO
    }
}

File: Sources/Chuck/main/model/norrischart/MapChartInPlaceUpdater.js
/*
 * Name: {MapChartInPlaceUpdater.js}
 * Module: {Model/NorrisChart}
 * Location: {Chuck/Main/Model/NorrisChart}
 * Date: {2015-04-12}
 * Version: {v0.6}
 *
 * History:
 *
 * ================================================================================
 * Version Date Programmer Changes
 * ================================================================================
 * 0.06 2015-05-21 Bigarella Chiara   Verify
 * ================================================================================
 * 0.05 2015-04-25 Dal Bianco Davide   Edit
 * ================================================================================
 * 0.04 2015-04-27 Carlon Chiara   Verify
 * ================================================================================
 * 0.03 2015-04-25 Pavanello Fabio Matteo   Edit
 * ================================================================================
 * 0.02 2015-04-14 Pavanello Fabio Matteo   Verify
 * ================================================================================
 * 0.01 2015-04-12 Moretto Alessandro   Creation
 * ================================================================================
 */

angular.module('norris-chartupdater')

.factory('MapChartInPlaceUpdater', [function () {
	/**
	 * Creates an instance of a new MapChartInPlaceUpdater, or returns the existing instance, if it already exits.
	 * @constructor
	 */
    function MapChartInPlaceUpdater() {
        if(!(this instanceof MapChartInPlaceUpdater)) return new MapChartInPlaceUpdater();
    }

    MapChartInPlaceUpdater.prototype.instance=new MapChartInPlaceUpdater(); // static
	
	/**
     * Returns the unique existing instance of the MapChartInPlaceUpdater
     * @return {MapChartInPlaceUpdater} - the unique existing instance of the class
     */
    MapChartInPlaceUpdater.getInstance = function() { // static
        return MapChartInPlaceUpdater.prototype.instance;
    };

    /**
     * Updates a map chart with in place method. The map chart data should not be empty.
     * @param {MapChartImpl} chart - the map chart to update;
     * @param updateData - the updating.
     */
    MapChartInPlaceUpdater.prototype.update = function (chart, updateData) {
        var isEmpty=function(obj) {
            for(var prop in obj) {
                if(obj.hasOwnProperty(prop))
                    return false;
            }
            return true;
        };

        if (!isEmpty(updateData)) {
            var data=chart.getData();
            if (!isEmpty(data)) {
                for(var i=0; i<updateData.length; i++) {
                    var series=updateData[i].position.series;
                    var index=updateData[i].position.index;
                    data[series].values[index].x=updateData[i].data.x;
                    data[series].values[index].y=updateData[i].data.y;
                }
                chart.setData(data);
            }
            else {
                console.log("ERROR: the chart has no data to update.");
                throw ("emptyChart");
            }
        }
    };

    return MapChartInPlaceUpdater;

}]);